Modernizations

dnicolodi opened the pull request.

dnicolodi committed:
Remove setup.py and supporting files

Keep meson-python as the only supported build system. Add an hack to
the Makefile to support in-place build for backward compatibility with
established development workflows.

dnicolodi committed:
Remove outdated targets from Makefile

dnicolodi committed:
Build and run C code test with Meson

dnicolodi committed:
ci: Run tests on Python 3.13 too

dnicolodi committed:
Adjust tests to Python 3.13

Python 3.13 strips common leading space from docstrings. However, the
algorithm used is not the same as the one used by textwrap.dedent().
This affects tests that use the test function docstring as input data.

blais commented on Makefile:
```diff
@@ -40,28 +29,16 @@ $(CROOT)/lexer.c $(CROOT)/lexer.h: $(CROOT)/lexer.l $(CROOT)/grammar.h
-
-$(CROOT)/tokens_test: $(CROOT)/tokens_test.o $(CROOT)/tokens.o $(CROOT)/decimal.o
+build:
+	meson setup --reconfigure build/
```
I'm all for modernization, but this does include a new dependency on meson though.
How reliable is that? Will it work on Windows? On macOS?

dnicolodi replied:
Yes, it works on Windows and macOS. It is how the binary Python packages we publish are built and also how the package is build if someone installs it on a platform for which we don't publish a binary. I think it is more reliable than setuptools. The build system we use is the same used by numpy, scipy, pandas, matplotlib and others.

dnicolodi commented:
Converting this as a draft for now. The Makefile does not regenerate the parser from the bison and flex sources anymore when running `make build`, therefore working on the parser is more cumbersome now. I would remove the generated sources from version control and have them generated at build time. We publish binary packages for the major platforms, thus it should not be a problem for users.

dnicolodi committed:
ci: Show compilation log for test jobs

blais commented:
Another thing to modernize Beancount would be to make it use `uv`

dnicolodi commented:
Use `uv` for doing what? You can build and install Beancount via `uv` just like any other Python package.

dnicolodi committed:
ci: Install flex and bison

dnicolodi committed:
Build lexer and grammar from flex and bison sources at build time

Remove the generated source code from the repository. This makes flex
and bison build dependencies. However, binary packages are provided
for all the major platforms (especially the ones where compiling from
source and installing these tools is a pain) thus the change should
have minimal impact on users.

dnicolodi committed:
ci: Run C code tests

dnicolodi committed:
Remove remnant of migration from Mercurial to Git

dnicolodi committed:
ci: Streamline and parallelize whells building

dnicolodi committed:
ci: Update cibuildwheel to version 2.21

dnicolodi committed:
ci: Install prerequisites to build wheels

dnicolodi committed:
ci: Build wheels for arm64 Windows too

dnicolodi commented:
@blais please have a look. I think this is ready to be merged now.

blais commented:
Changes look good.
I want to test those out on my machine later today before merging.

I'm not sure however that it's the best idea to require flex/bison just to build; I was checking in the C code on purpose to avoid those dependencies. We can try it, but if it prevents too many people from installing I'd consider bringing it back in the future.

More later.

dnicolodi commented:
I would be surprised to find that anyone installing Beancount is building from source, unless they want an unreleased version. We provide binary packages for all major platforms. The only platform that supports building from from source without too many problems is Linux and even there someone would need to install a C compiler and the Python development package. Adding flex and bison on top does not seem much of a problem to me. I don't seem to find a way to look at statistics of dowloaded source distributions vs binary distributions, but I think there must be a way to get at it.

blais committed:
Added missing -lm to tests (Linux PopOS/Ubuntun 22.04).

blais commented:
Ok, this works on my box. Looks elegant as well.

blais commented:
One more thing: my logic for baking the revision into the source code is gone it seems.
This was useful when developing and using git commands to detect a bad version.
I'm referring to VC_CHANGESET and VC_TIMESTAMP.
Not super critical, but I wonder if it'd be easy to reinstall.
I guess with Windows it would be difficult to run git as an external program from meson.build.

dnicolodi commented:
Some of the logic is still there https://github.com/beancount/beancount/blob/af2ba01e71865a76d42bf70b0fe7519cd9a5f087/meson.build#L203-L214
however I did not test it. The one setting `VC_CHANGESET` and `VC_TIMESTAMP` is gone indeed. I don't remember why I didn't implement it. Most likely because it does not work when building from a source distribution and because the Bazel build also dropped support for it. I can restore it.

blais commented:
Nah that's alright, don't worry about it. If it gets annoying - e.g. we end up often using the wrong build - then we'll build it again. Let's leave the ifdefs for now they're harmless. 
tyty Daniele,
